larpandoracontent/LArHelpers/LArGeometryHelper.cc

```diff
@@ -422,6 +422,74 @@ bool LArGeometryHelper::IsInGap3D(const Pandora &pandora, const CartesianVector
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+float LArGeometryHelper::ProjectAcrossGap3D(const Pandora &pandora, const CartesianVector &testPoint, const CartesianVector &testDirection,
+        const float gapTolerance, const int recurseLimit)
+{
+    float bestGapSize = std::numeric_limits<float>::max();
+
+    for (const DetectorGap *const detectorGap : pandora.GetGeometry()->GetDetectorGapList())
+    {
+        if (!detectorGap->IsInGap(testPoint, TPC_3D, gapTolerance))
+            continue;
+
+        const LineGap* lineGap = dynamic_cast<const LineGap *>(detectorGap);
+
+        if (lineGap == nullptr)
+            continue;
+
+        // ATTN: Right now, this only works for gaps with only X components.
+        //       If the Z is used for detector gaps, can cause infs.
+
+        // INFO: Get the angles from the current point to each gap edge.
+        const CartesianVector startPoint(lineGap->GetLineStartX(), testPoint.GetY(), testPoint.GetZ());
+        const float distanceToStart = std::fabs(testPoint.GetX() - lineGap->GetLineStartX());
+        const float angleToStart = startPoint.GetCosOpeningAngle(testDirection);
+
+        const CartesianVector endPoint(lineGap->GetLineEndX(), testPoint.GetY(), testPoint.GetZ());
+        const float distanceToEnd = std::fabs(testPoint.GetX() - lineGap->GetLineEndX());
+        const float angleToEnd = endPoint.GetCosOpeningAngle(testDirection);
+
+        // INFO: Choose the gap infront of the current point.
+        const bool shouldUseGapStart = angleToStart > angleToEnd ? true : false;
```

Comment:
Why not using directly the distance from the edges, to determine whether to use the start or the end of the gap? I may be missing something, but it seems to me that angleToStart and angleToEnd do not depend on the testPoint x coordinate...